compiler/src/iree/compiler/Codegen/LLVMGPU/VectorReductionToGPU.cpp

```diff
@@ -0,0 +1,186 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "iree/compiler/Codegen/Utils/Utils.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Vector/Transforms/VectorDistribution.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+/// Emit shared local memory allocation in case it is needed when lowering the
+/// warp operations.
+static Value allocateGlobalSharedMemory(Location loc, OpBuilder &builder,
+                                        vector::WarpExecuteOnLane0Op warpOp,
+                                        Type type) {
+  builder.setInsertionPoint(warpOp);
```

Comment:
This insertionPoint does not play nicely with the caller and may result in use-after-def.

This IR would fail to lower to scf:
```
func.func @reduce_dispatch_0() {
  %cst = arith.constant dense<1.000000e+00> : vector<128xf32>
  %c0 = arith.constant 0 : index
  %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) offset(%c0) alignment(64) : memref<128x384xf32>
  memref.assume_alignment %0, 64 : memref<128x384xf32>
  %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) offset(%c0) alignment(64) : memref<128xf32>
  memref.assume_alignment %1, 64 : memref<128xf32>
  %2 = gpu.thread_id  x
  %3 = vector.warp_execute_on_lane_0(%2)[32] -> (vector<4xf32>) {
    vector.yield %cst : vector<128xf32>
  }
  %4 = affine.apply affine_map<()[s0] -> (s0 * 4)>()[%2]
  vector.transfer_write %3, %1[%4] {in_bounds = [true]} : vector<4xf32>, memref<128xf32>
  return
}
```

Reply:
Dropping the extra setInsertionPoint here is the way to go, the builder is already set at the proper insertion point by the caller.